File: js/assignment11/assignment11.test.js
import { expect, describe, it, beforeAll, vi } from "vitest";
import main from "./assignment11.js";

describe("main", () => {
   it("Should print the correct output to the console", () => {
      let output = "";
      const originalWrite = process.stdout.write;

      // Override process.stdout.write to capture the output
      process.stdout.write = (str) => {
         output += str;
         return true;
      };

      main();

      const expectedOutput = [
         "\tItem           Projected\n",
         "\t=============  ==========\n",
         "\tIncome         $  1000.00\n",
         "\tTaxes          $   100.00\n",
         "\tTithing        $   100.00\n",
         "\tLiving         $   650.00\n",
         "\tOther          $    90.00\n",
         "\t=============  ==========\n",
         "\tDelta          $    60.00\n"
      ].join("");

      // Restore the original process.stdout.write
      process.stdout.write = originalWrite;

      expect(output).toBe(expectedOutput);
   });
});

describe("key", () => {
   let key;
   beforeAll(async () => {
      try {
         key = await import("./etc/.key/assignment11.js").then(module => module.main);
      } catch (error) {
         console.warn("Key file not found. Skipping key tests.");
         key = null;
      }
   });

   if (!key) {
      it.skip("Key file not found. Skipping key tests.");
   }
   else {
      it("Should print the correct output to the console", () => {
         let output = "";
         const originalWrite = process.stdout.write;

         // Override process.stdout.write to capture the output
         process.stdout.write = (str) => {
            output += str;
            return true;
         };

         key();

         const expectedOutput = [
            "\tItem           Projected\n",
            "\t=============  ==========\n",
            "\tIncome         $  1000.00\n",
            "\tTaxes          $   100.00\n",
            "\tTithing        $   100.00\n",
            "\tLiving         $   650.00\n",
            "\tOther          $    90.00\n",
            "\t=============  ==========\n",
            "\tDelta          $    60.00\n"
         ].join("");

         // Restore the original process.stdout.write
         process.stdout.write = originalWrite;

         expect(output).toBe(expectedOutput);
      });
   }
});

File: python/assignment11/assignment11_test.py
"""
Name: assignment11_test.py
Description:
    This file will be collected by the pytest command and test assignment11.py
Author:
    Elijah Harrison
"""
import io
import sys
import pytest
from assignment11 import main

try:
    from etc.key.assignment11_proxy import main as key
except Exception as e:
    key = None
    print(f"Could not import key from path etc.key.assignment11_proxy: {e}")

"""
Import Statements/Globals
"""
expected_output = (
    "\tItem           Projected\n"
    "\t=============  ==========\n"
    "\tIncome         $  1000.00\n"
    "\tTaxes          $   100.00\n"
    "\tTithing        $   100.00\n"
    "\tLiving         $   650.00\n"
    "\tOther          $    90.00\n"
    "\t=============  ==========\n"
    "\tDelta          $    60.00\n"
)

def test_main(capsys):
    # Call the function
    main()

    # Capture the output
    captured = capsys.readouterr()

    # Assert the output
    assert captured.out == expected_output

def test_key(capsys):
    """
    For Mr. Eli's key.
    """
    if not key:
        pytest.skip("Skipping test_key because key is not available")
    
    # Call the function
    key()

    # Capture the output
    captured = capsys.readouterr()

    # Assert the output
    assert captured.out == expected_output
